recipes/generate.py

```diff
@@ -0,0 +1,176 @@
+        self._model = self._setup_model(
+            model_cfg=cfg.model,
+            model_state_dict=ckpt_dict[utils.MODEL_KEY],
+        )
+        self._tokenizer = config.instantiate(cfg.tokenizer)
+
+    def _setup_model(
+        self,
+        model_cfg: DictConfig,
+        model_state_dict: Dict[str, Any],
+    ) -> nn.Module:
+        with utils.set_default_dtype(self._dtype), self._device:
+            model = config.instantiate(model_cfg)
+
+        model.load_state_dict(model_state_dict)
+
+        # Validate model was loaded in with the expected dtype.
+        utils.validate_expected_param_dtype(model.named_parameters(), dtype=self._dtype)
+        logger.info(f"Model is initialized with precision {self._dtype}.")
+
+        # Ensure the cache is setup on the right device
+        with self._device:
+            model.setup_caches(max_batch_size=1, dtype=self._dtype)
+        return model
+
+    def _multinomial_sample_one_no_sync(self, probs_sort):
+        q = torch.empty_like(probs_sort).exponential_(1)
+        return torch.argmax(probs_sort / q, dim=-1, keepdim=True).to(dtype=torch.int)
+
+    def _logits_to_probs(self, logits: torch.Tensor, temperature: float, top_k: int):
+        logits = logits / max(temperature, 1e-5)
+
+        if top_k is not None:
+            v, _ = torch.topk(logits, min(top_k, logits.size(-1)))
+            pivot = v.select(-1, -1).unsqueeze(-1)
+            logits = torch.where(logits < pivot, -float("Inf"), logits)
+        probs = torch.nn.functional.softmax(logits, dim=-1)
+        return probs
+
+    def _sample(self, logits: torch.Tensor, temperature: float, top_k: int):
+        probs = self._logits_to_probs(logits[0, -1], temperature, top_k)
+        idx_next = self._multinomial_sample_one_no_sync(probs)
+        return idx_next, probs
+
+    def _prefill(
+        self,
+        model: nn.Module,
+        x: torch.Tensor,
+        input_pos: torch.Tensor,
+        temperature: float,
+        top_k: int,
+    ) -> torch.Tensor:
+        # input_pos: [B, S]
+        logits = model(x, input_pos)
+        return self._sample(logits, temperature, top_k)[0]
+
+    def _decode_one_token(
+        self,
+        model: nn.Module,
+        x: torch.Tensor,
+        input_pos: torch.Tensor,
+        temperature: float,
+        top_k: int,
+    ):
+        # input_pos: [B, 1]
+        assert input_pos.shape[-1] == 1
+        logits = model(x, input_pos)
+        return self._sample(logits, temperature, top_k)
+
+    def _decode_n_tokens(
+        self,
+        model: nn.Module,
+        cur_token: torch.Tensor,
+        input_pos: torch.Tensor,
+        num_new_tokens: int,
+        temperature: float,
+        top_k: int,
+    ):
+        new_tokens, new_probs = [], []
+        for i in range(num_new_tokens):
+            with torch.backends.cuda.sdp_kernel(
+                enable_flash=False, enable_mem_efficient=False, enable_math=True
+            ):
```

Comment:
I assume this is for performance reasons? Might be good to add a bit of detail on why we're doing this though (is it because the math implementation is faster when seq len and batch size are small?)